Deprecate legacy moderation system

laymonage opened the pull request:
<!-- Thanks for contributing to Wagtail! 🎉  Please add a description below, explaining the purpose of this pull request - including the issue number of the issue you're fixing (if applicable). -->

Ref: https://docs.wagtail.org/en/stable/releases/2.10.html#move-to-new-configurable-moderation-system-workflow

Workflows have been around since 2.10 and the legacy moderation system (one-step moderation) has been pretty much unused since. Per the release notes, you can't submit pages to the legacy system anymore, and the code is just kept to let people complete the moderations that were still ongoing. It's about time we remove the code for the legacy moderation system. Before doing so, ensure we raise enough warning so people are aware about this and won't have any page that are still in the old moderation system.

After the deprecation, we'll remove all code around the warning. We'll also remove the `submitted_for_moderation` field from the `Revision` model, along with other places where it's unavoidably referenced (e.g. when saving/publishing revisions).

There are places where I didn't raise a `RemovedInWagtail60Warning` explicitly, as I'm afraid that would trigger warnings even when there are no revisions in the old moderation system. I've marked them with comments though, so we can easily find it when we prepare Wagtail 6.0.

Searching the codebase for `moderat`, I think this is everything, but might be worth doing a double check in case I missed anything.


_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [x] For new features: Has the documentation been updated accordingly?

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

lb- commented:
Will this address https://github.com/wagtail/wagtail/issues/9106 ?

Or will we just leave that issue out of scope? Note Thibaud's comments on that issue also, maybe we just won't add these buttons back.

laymonage commented:
@lb- I did consider it but I think it's out of scope. With workflows, developers can have custom Tasks with custom forms, which means that approving/rejecting a task can be more complex than a click of a button. For example, the form may have required fields, so this means the userbar item will need to support showing the form to the user and submitting the data as well. Not saying it's impossible, especially with the `SwapController` in place, but it would require a significant amount of work.

jacobtoppm reviewed (commented):
Generally, this is all looking good. However, wondering whether we should provide some guidance on moving - maybe a "here is how you see all your pages in the old system" to approve/reject them, or even a management command to auto-resubmit pages to the new workflow system (though tbh, given that the new system works on page rather than revision level, that command's implementation could cause conusion of its own)

laymonage commented:
> Generally, this is all looking good. However, wondering whether we should provide some guidance on moving - maybe a "here is how you see all your pages in the old system" to approve/reject them,

I was thinking of adding that but I wasn't sure of a few things:
- Is there a possibility that a page has been operating normally with workflows etc., but there are stale revisions that have `submitted_for_moderation=True`? I'm guessing no (unless the revision is created programmatically and incorrectly handled from that point onwards), but just wanted to make sure
- Is querying for `Revisions` with `submitted_for_moderation=True` and calling `approve_moderation()`/`reject_moderation()` enough?

> or even a management command to auto-resubmit pages to the new workflow system (though tbh, given that the new system works on page rather than revision level, that command's implementation could cause conusion of its own)

I don't think I can pull it off, as I don't have a clear idea on how the legacy system worked (and any of its nuances), much less properly writing the migration to workflows. It's probably also not worth it given that it's been more than three years and we've gone through so many releases since we stopped supporting it for newly submitted pages.

jacobtoppm commented:
Yeah, I agree (as in my comment above) that the management command is probably overkill!

It's been a while since I touched the old moderation system, but:
"Is there a possibility that a page has been operating normally with workflows etc., but there are stale revisions that have submitted_for_moderation=True?" -> Needs testing, I think, but I believe it's a distant possibility.
"Is querying for Revisions with submitted_for_moderation=True and calling approve_moderation()/reject_moderation() enough?" -> I think so. 

Given it's such an old system, it's probably sufficient to say:
- How a user in the admin would see all pages in the legacy system + approve/reject
- How someone could programatically reject (or less likely, approve) all pages in the legacy system

gasman commented:
I always assumed that the reason we didn't provide a migration path when the new workflows system was introduced, was that it was considered unfeasible at the time. I don't see any reason to revisit that decision now, especially as the number of people who would benefit from it now is so much lower.

I think we're fine to just note that the system is being removed, and advise people to clear their moderation backlog before upgrading - if there are factors that prevent them from doing that, that's their own problem.

jacobtoppm commented:
Yeah, it's a pretty different system.

That's fair. I just think the note should point people at where you actually see your backlog in order to clear it, as a reminder.

laymonage commented on docs/releases/5.2.md:
```diff
@@ -154,6 +154,31 @@ def register_frank_menu_item():
+
+The legacy moderation system, which was replaced by the new workflow system in Wagtail 2.10, is now deprecated. Since Wagtail 2.10, submitting a page for moderation will use the new workflow system. However, the legacy moderation system is still in place for approving and rejecting pages that were submitted for moderation before Wagtail 2.10.
+
+To view all pages that are still in the legacy moderation system backlog, you can sign in as a superuser and see if there is a "Pages awaiting moderation" section in the admin dashboard. You can approve or reject the pages from there. You can also do this programmatically by querying for `Revision.objects.filter(submitted_for_moderation=True)` and calling `revision.approve_moderation()` or `revis ...
```
I added this paragraph @jacobtoppm, I believe it's enough?

I tried saving a new revision with `submitted_for_moderation=True`, which made the page show on the dashboard and I tried approving/rejecting from there. I also tried doing it programmatically and it seems to work.

jacobtoppm replied:
Yep, that's enough for me! I suspect the likely crossover is very small, but I think it's nicer to have a little pointer like this

lb- commented:
Thanks @laymonage - this is great to see.

laymonage commented:
Thanks all, I noticed a bug in adding the warning to `Revision.submitted_revisions`: I shouldn't have used `cached_property` because managers can only be accessed from the class (not the instance). Fixed it by using `classproperty` instead and I added tests to make sure it works as expected. I also noticed using `stacklevel=1` (the default) instead of `stacklevel=2` when raising the warning produces more helpful trace in the output, so I changed it too.

Changes are in https://github.com/wagtail/wagtail/compare/08fdb4c206a02e64bd42d6c4d36942a191ac3f78..fcfecd5f02f577a9e948f5ff0ef5126cb3616670

Had to [fix the test](https://github.com/wagtail/wagtail/compare/fcfecd5f02f577a9e948f5ff0ef5126cb3616670..d0361e98296bb59d99b4089b0c2a0e84cf9cdf3d) because I was using `self.assertQuerySetEqual` which is only available in Django 4.2+, we added the `self.assertQuerysetEqual()` in `WagtailTestUtils` but the test class in this one doesn't inherit from that so I just used plain `assertEqual()` with `list()`.
